cdc/sink/codec/craft/encoder.go

```diff
@@ -0,0 +1,321 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.orglicensesLICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package craft
+
+import (
+	"encoding/binary"
+	"math"
+	"unsafe"
+
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/ticdc/cdc/model"
+)
+
+// create byte slice from string without copying
+func unsafeStringToBytes(s string) []byte {
+	return *(*[]byte)(unsafe.Pointer(
+		&struct {
+			string
+			Cap int
+		}{s, len(s)},
+	))
+}
+
+// Primitive type encoders
+func encodeFloat64(bits []byte, data float64) []byte {
+	v := math.Float64bits(data)
+	return append(bits, byte(v), byte(v>>8), byte(v>>16), byte(v>>24), byte(v>>32), byte(v>>40), byte(v>>48), byte(v>>56))
+}
+
+func encodeVarint(bits []byte, data int64) []byte {
```

Comment:
What about using`encoding/binary.PutVarint`?

Reply:
It's a small optimization. They actually does the same thing except that the implementation here does not require slice to have binary.MaxVarintLen64 bytes available. So the slice can be used without resizing after encoding.